### PR TITLE
🎨 Use Pipeline for processing Capsules

### PR DESCRIPTION
**Pull request by andersjohnsen**

### CODE REVIEW

File: pkg/controller/pipeline.go [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+type ObjectKey struct {
+	client.ObjectKey
+	schema.GroupVersionKind
+}
+
+func (ok ObjectKey) String() string {
+	return fmt.Sprintf("%s/%s", ok.GroupVersionKind.String(), ok.ObjectKey.String())
+}
+
+func (ok ObjectKey) MarshalLog() interface{} {
```
Comment (MatiasFrank) [inline pkg/controller/pipeline.go:75]: Does this need to return `interface{}`? It seems the return value is statically known

File: pkg/controller/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+	}{
+		Name:      ok.Name,
+		Namespace: ok.Namespace,
+		Group:     ok.Group,
+		Kind:      ok.Kind,
+		Version:   ok.Version,
+	}
+}
+
+func Equal(a ObjectKey, b ObjectKey) bool {
```
Comment (MatiasFrank) [inline pkg/controller/pipeline.go:91]: Do we need a function for this?

File: pkg/controller/pipeline.go [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+func (p *Pipeline) Config() *configv1alpha1.OperatorConfig {
+	return p.config
+}
+
+func (p *Pipeline) Scheme() *runtime.Scheme {
+	return p.scheme
+}
+
+func (p *Pipeline) Capsule() *v1alpha2.Capsule {
+	return p.capsule
```
Comment (MatiasFrank) [inline pkg/controller/pipeline.go:157]: Should we return deepcopies here (and for `Config` to ensure plugins aren't naughty and modifies stuff?

File: pkg/controller/pipeline.go (hunk 1 above)
Comment (andersjohnsen) [reply pkg/controller/pipeline.go:75]: It's matching an interface to generate log messages, so not something we dictate

File: pkg/controller/pipeline.go (hunk 2 above)
Comment (andersjohnsen) [reply pkg/controller/pipeline.go:91]: removed

File: pkg/controller/pipeline.go (hunk 3 above)
Comment (andersjohnsen) [reply pkg/controller/pipeline.go:157]: Sure, will do

File: pkg/controller/pipeline.go [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+			}
+		}
+
+		p.logger.Info("run steps", "current_objects", maps.Keys(p.currentObjects))
+
+		for _, s := range p.steps {
+			if err := s.Apply(ctx, p); err != nil {
+				if err := p.updateStatus(ctx, nil, err); err != nil {
+					return err
+				}
```
Comment (MatiasFrank) [inline pkg/controller/pipeline.go:274]: A couple times you use the pattern of calling `updateStatus` when we hit an error and then returning the original error.
We can wrap this entire loop in a function call, then call `updateStatus` afterwards if the function returns err. Then updating status on error is easier to maintain

Comment (andersjohnsen) [reply pkg/controller/pipeline.go:274]: Done!

**Pull request merged**